bundles/org.eclipse.equinox.preferences/src/org/eclipse/core/internal/preferences/PreferencesService.java

```diff
@@ -149,8 +144,12 @@ public IStatus applyPreferences(IExportedPreferences preferences) throws CoreExc
 						}
 					}
 				}
+				if (!propsToRemove.isEmpty() && !(globalNode instanceof EclipsePreferences)) {
+					// intern strings we import because some people in their property change
+					// listeners use identity instead of equals. See bug 20193 and 20534.
+					propsToRemove.replaceAll(String::intern);
+				}
```

Comment:
The elements in the keys array above are obtained from an `ExportedPreferences`, which is a subclass of `EclipsePreference` and therefore has its keys already interned (unless somebody extends this internl class and changes the method to store elements, but those that do that should take care of interning themself).

On the other hand `globalNode` is just a `IEclipsePreferences` node from the tree and because `Clients may implement this interface`, it can be an arbitrary implementation. Therefore interning  `propsToRemove` can only be skipped if the node is of a `EclipsePreferences` subclass.

In general I would call the listeners that rely on string identity comparison badly implemented.